expression/collation.go

```diff
@@ -189,12 +168,103 @@ func deriveCoercibilityForColumn(c *Column) Coercibility {
 	return CoercibilityImplicit
 }
 
+func deriveCollation(ctx sessionctx.Context, funcName string, args []Expression, retType types.EvalType, argTps ...types.EvalType) (dstCharset, dstCollation string, coercibility Coercibility, err error) {
+	switch funcName {
+	case ast.Concat, ast.ConcatWS, ast.Lower, ast.Reverse, ast.Upper, ast.Quote, ast.Coalesce:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args...)
+	case ast.Left, ast.Right, ast.Repeat, ast.Trim, ast.LTrim, ast.RTrim, ast.Substr, ast.SubstringIndex, ast.Replace:
```

Comment:
Why doesn't `REPLACE` check all arguments？

By the way, are these checked function by function?

Reply:
MySQL does this. I think it is because the result of this function is based on the first argument, it does not suppose to change to another charset.

> are these checked function by function

yes, the listed function is checked function by function